src/util/MathUtil.cpp

```diff
@@ -16,6 +18,189 @@ bool MathUtil::AlmostEqual(double x, double y, int precision) {
   return abs(x - y) < 5*pow(10.0, -(precision + 1));
 }
 
+// added by Yang
+int MathUtil::factorial(int n) {
+  int product = 1;
+  for (int i = 1; i <= n; i++) { product *= i; }
+  return product;
+}
+
+// directly copied from CKronik2.cpp
+double MathUtil::gammaln(double xx) {
+  double x,y,tmp,ser;
+  static double cof[6]={76.18009172947146,-86.50532032941677,24.01409824083091,-1.231739572450155,0.1208650973866179e-2,-0.5395239384953e-5};
+  int j;
+
+  y=x=xx;
+  tmp=x+5.5;
+  tmp-=(x+0.5)*log(tmp);
+  ser=1.000000000190015;
+  for(j=0;j<=5;j++) ser+=cof[j]/++y;
+  return -tmp+log(2.5066282746310005*ser/x);
+}
+
+double MathUtil::LogNChooseK(int n, int k) {
+	if (n <= 0 || n < k) { return 0; }
+	double result = 0;
+	for (int i = 1; i <= k; i++) { result += log(n-i+1) - log(i); }
+	return result;
+}
+
+double MathUtil::LogSumExp(std::vector<double>* log_values) {
+	if (log_values->size() <= 0) { return 0.0; }
+	else {
+		double max_value = *max_element(log_values->begin(), log_values->end());
+		double sum = 0;
+		for (int i = 0; i < log_values->size() ; i++){ sum += exp(log_values->at(i) - max_value); }
+		return log(sum) + max_value;
+	}
+}
+
+double MathUtil::MaxInArr(double* arr_values, int size) {
+	if (arr_values == NULL || size <= 0) { return 0; }
+
+	double max_value = arr_values[0];
+	for (int i=0; i<size; ++i) {
+		if (arr_values[i] > max_value) { max_value = arr_values[i]; }
+	}
+	return max_value;
+}
+
+double MathUtil::NormalizedDotProduct(double* src_values, double* tgt_values, int size, bool take_sqrt) {
+	if (src_values == NULL || tgt_values == NULL || size <= 0) { return 0; }
+
+	double prod_sum = 0, src_sum = 0, tgt_sum = 0;
+	for (int i=0; i<size; ++i) {
+		if (take_sqrt) {
+			prod_sum += sqrt(src_values[i]) * sqrt(tgt_values[i]);
+			src_sum += src_values[i];
+			tgt_sum += tgt_values[i];
+		} else {
+			prod_sum += src_values[i] * tgt_values[i];
+			src_sum += src_values[i] * src_values[i];
+			tgt_sum += tgt_values[i] * tgt_values[i];
+		}
+	}
+
+	if (AlmostEqual(prod_sum, 0, 4)) { return 0; }
+
+	double result = prod_sum / sqrt(src_sum * tgt_sum);
+	return (result == result) ? result : 0; // this is to avoid nan
+}
+
+std::vector<double> MathUtil::linspace(double start, double end, int num) {
+	std::vector<double> linspaced;
+
+	if (num < 2) { throw runtime_error("linspace should contain at least two points"); }
+	double delta = (end - start) / (num - 1);
+	for(int i=0; i < num-1; ++i) { linspaced.push_back(start + delta * i); }
+	linspaced.push_back(end);
+
+	return linspaced;
+}
+
+int MathUtil::binarySearch(const double* data_arr, int data_size, double query) {
```

Comment:
There is also a std vesion of binary search....why did you reimplement?

Reply:
std function is based on the vector instead of array